velox/exec/tests/utils/AggregationFuzzer.cpp

```diff
@@ -778,7 +791,14 @@ void AggregationFuzzer::go() {
           }
         }
 
-        auto input = generateInputData(argNames, argTypes);
+        // 10% of times test global aggregation with empty input.
```

Comment:
I feel we should go with a different approach (instead of choosing empty input 10% of times for global agg). We want to add a plan for global aggregation with empty input here: https://github.com/facebookincubator/velox/blob/main/velox/exec/tests/utils/AggregationFuzzer.cpp#L874

Reply:
Thanks for the suggestion, I tried to add a plan here first, but that threw the following error:
```
facebook::velox::exec::test::PlanBuilder::fields(const std::vector<…> &) PlanBuilder.cpp:1667
facebook::velox::exec::test::PlanBuilder::aggregation(const std::vector<…> &, const std::vector<…> &, const std::vector<…> &, const std::vector<…> &, facebook::velox::core::AggregationNode::Step, bool, const std::vector<…> &) PlanBuilder.cpp:723
facebook::velox::exec::test::PlanBuilder::aggregation(const std::vector<…> &, const std::vector<…> &, const std::vector<…> &, const std::vector<…> &, facebook::velox::core::AggregationNode::Step, bool) PlanBuilder.h:404
facebook::velox::exec::test::PlanBuilder::partialAggregation(const std::vector<…> &, const std::vector<…> &, const std::vector<…> &) PlanBuilder.h:295
facebook::velox::exec::test::makeAlternativePlansWithValues(const std::vector<…> &, const std::vector<…> &, const std::vector<…> &, const std::vector<…> &, const std::vector<…> &, std::vector<…> &) AggregationFuzzer.cpp:919
facebook::velox::exec::test::AggregationFuzzer::verifyAggregation(const std::vector<…> &, const std::vector<…> &, const std::vector<…> &, const std::vector<…> &, bool, const std::vector<…> &) AggregationFuzzer.cpp:1261
facebook::velox::exec::test::AggregationFuzzer::go() AggregationFuzzer.cpp:735
facebook::velox::exec::test::aggregateFuzzer(std::unordered_map<…>, unsigned long, const std::unordered_map<…> &, facebook::velox::VectorFuzzer::Options::TimestampPrecision, const std::unordered_map<…> &, const std::optional<…> &, std::unique_ptr<…>) AggregationFuzzer.cpp:334
facebook::velox::exec::test::AggregationFuzzerRunner::runFuzzer(unsigned long, const std::optional<…> &, std::unique_ptr<…>, const facebook::velox::exec::test::AggregationFuzzerRunner::Options &) AggregationFuzzerRunner.h:143
facebook::velox::exec::test::AggregationFuzzerRunner::run(unsigned long, std::unique_ptr<…>, const facebook::velox::exec::test::AggregationFuzzerRunner::Options &) AggregationFuzzerRunner.h:108
main AggregationFuzzerTest.cpp:121
start 0x0000000189bdd058
```

@Krishna-Prasad-P-V could you please help look into this?

Reply:
@pramodsatya What is the command used to run aggregation fuzzer locally?

Reply:
I have a fix to pass outputType from plan in `verifyAggregation` method via https://github.com/facebookincubator/velox/blob/c5d76db8516efbaa9ba95b6faa02aac8c5aff446/velox/exec/tests/utils/AggregationFuzzer.cpp, but a repro would be useful to confirm.

Reply:
@Krishna-Prasad-P-V you can use the following command to run aggregation fuzzer locally:
```
./velox_aggregation_fuzzer --seed 2132047877 --duration_sec 60 --logtostderr=1 --minloglevel=0 --repro_persist_path=/tmp/aggregate_fuzzer_repro
```

Reply:
I am getting this error while running aggregation fuzzer.
```
I1018 23:30:11.312508 442994 AggregationFuzzer.cpp:722] ==============================> Started iteration 13 (seed: 1959933215)
*** Aborted at 1697697011 (Unix time, try 'date -d @1697697011') ***
*** Signal 8 (SIGFPE) (0x557e6dd79f6c) received by PID 442994 (pthread TID 0x7f0c08f26400) (linux TID 442994) (code: integer divide by zero), stack trace: ***
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(_ZN5folly10symbolizer17getStackTraceSafeEPmm+0x72)[0x557e7434bdd2]
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(_ZN5folly10symbolizer21SafeStackTracePrinter15printStackTraceEb+0x3a)[0x557e7434d7fa]
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(+0xed76751)[0x557e7434b751]
/lib/x86_64-linux-gnu/libc.so.6(+0x4251f)[0x7f0c0b0bb51f]
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(_ZN8facebook5velox12VectorFuzzer19fuzzOffsetsAndSizesERN5boost13intrusive_ptrINS0_6BufferEEES6_mm+0xe6)[0x557e6dd79f6c]
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(_ZN8facebook5velox12VectorFuzzer9fuzzArrayERKSt10shared_ptrINS0_10BaseVectorEEi+0x6e)[0x557e6dd7a0d6]
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(_ZN8facebook5velox12VectorFuzzer11fuzzComplexERKSt10shared_ptrIKNS0_4TypeEEi+0x189)[0x557e6dd798c3]
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(_ZN8facebook5velox12VectorFuzzer4fuzzERKSt10shared_ptrIKNS0_4TypeEEi+0x1d7)[0x557e6dd77c2b]
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(_ZN8facebook5velox12VectorFuzzer7fuzzRowERKSt10shared_ptrIKNS0_7RowTypeEEib+0x7d)[0x557e6dd7b045]
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(+0x7982bdf)[0x557e6cf57bdf]
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(+0x79843bf)[0x557e6cf593bf]
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(_ZN8facebook5velox4exec4test15aggregateFuzzerESt13unordered_mapINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt6vectorISt10shared_ptrINS1_26AggregateFunctionSignatureEESaISD_EESt4hashIS9_ESt8equal_toIS9_ESaISt4pairIKS9_SF_EEEmRKS3_IS9_S9_SH_SJ_SaISK_ISL_S9_EEENS0_12VectorFuzzer7Options18TimestampPrecisionEST_RKSt8optionalIS9_ESt10unique_ptrINS2_20ReferenceQueryRunnerESt14default_deleteIS12_EE+0x15d)[0x557e6cf54704]
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(_ZN8facebook5velox4exec4test23AggregationFuzzerRunner9runFuzzerEmRKSt8optionalINSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEEESt10unique_ptrINS2_20ReferenceQueryRunnerESt14default_deleteISF_EERKNS3_7OptionsE+0x1e4)[0x557e6cf3d96c]
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(_ZN8facebook5velox4exec4test23AggregationFuzzerRunner3runEmSt10unique_ptrINS2_20ReferenceQueryRunnerESt14default_deleteIS5_EERKNS3_7OptionsE+0x69)[0x557e6cf3d513]
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(main+0x8d9)[0x557e6cf3b0e1]
/lib/x86_64-linux-gnu/libc.so.6(+0x29d8f)[0x7f0c0b0a2d8f]
/lib/x86_64-linux-gnu/libc.so.6(__libc_start_main+0x7f)[0x7f0c0b0a2e3f]
./debug/velox/exec/tests/velox_aggregation_fuzzer_test(_start+0x24)[0x557e6cf384b4]
(safe mode, symbolizer not available)
Floating point exception (core dumped)
```

Reply:
Could you push your changes to this PR please? Will take a look.

Reply:
Hi @aditi-pandit, moving the plan with empty input and global aggregation to the function `makeAlternativePlansWithValues` causes the result computed with this plan to be different from the other plans generated in this function. The aggregation fuzzer fails as a result since there is a check to ensure the output from each plan is the same.
Is the current approach fine? Or should we introduce a similar `makePlans` function to generate a set of equivalent plans for empty input without group by?